### ISSUE TITLE
ExoPlayer freezes on STATE_BUFFERING if connection has changed

### ISSUE THREAD

**Issue by don11995** [google/ExoPlayer#911]
I'm writing a portable player for android devices and I found interesting bug: if connection has changed (for example 3G changed to 2G or 3G changed to WiFi) player freezes on STATE_BUFFERING and do not anithing untill player is reinitialized. Call pause and than play, in this case, not work. I resolved this problem by adding timer to callback: if STATE_BUFFERING is more than 20 seconds - init player again. I'm interested in this project, that's why I'm writting any problems that I find here. Thanks.

**Comment by ojw28** [comment]
This most likely has the same root cause as #910, if it's the same type of media. In any case, let's consolidate these issues under that single issue.

**Comment by oresot** [comment]
Hello!

I'm commenting here as the thread #910 is closed (and I can't find the relation). I'm having exactly the same issue. 

When connection changes from 3G to Wifi, or between different Wifi networks, ExoPlayer automatically stops playing (shouldn't it keep playing until it runs out of buffered data?) and moves to STATE_BUFFERING and getting stuck there. isReady always return false in this case.

This is the stream I'm using: http://188.165.222.163:8012 (audio stream in MP3)

I am preparing my ExoPlayer like this:

        private static final int BUFFER_SEGMENT_SIZE = 64 * 1024;

        private static final int BUFFER_SEGMENT_COUNT = 256;
        player = ExoPlayer.Factory.newInstance(1);
        playerControl = new PlayerControl(player);

        Uri uri = Uri.parse(url);

        String userAgent = Util.getUserAgent(this.context, "ExoPlayer");

        Allocator allocator = new DefaultAllocator(BUFFER_SEGMENT_SIZE);
        DataSource dataSource = new DefaultUriDataSource(this.context, null, userAgent);

        ExtractorSampleSource sampleSource = new ExtractorSampleSource(
                uri, dataSource, allocator, BUFFER_SEGMENT_COUNT * BUFFER_SEGMENT_SIZE);

        MediaCodecAudioTrackRenderer audioRenderer = new MediaCodecAudioTrackRenderer(sampleSource);
        player.prepare(audioRenderer);
        player.setPlayWhenReady(true);


Thanks!

**Comment by don11995** [comment]
@oresot Hello! Yes, my preparing is the same, except  DefaultUriDataSource - I'm using OkHttpDataSource, because I need to use ICYCAST protocol and DefaultUriDataSource does not support it correctly on different devices. I resolve this problem by adding listener and timer to player instance and checking STATE_BUFFERING: if buffering was longer than 30 sec - reinit player. Sorry, for my English.

**Comment by oresot** [comment]
@don11995 I had to do a trick like that too. My stream is from a Shoutcast server, do you know if that means its ICYCAST too? I didn't set up the streaming server.

**Comment by don11995** [comment]
@oresot I don't know what protocol Shoutcast provides, but if you are in doubt, use OkHttpExtension from ExoPlayer's extensions folder. It supports more protocols than DefaultUriDataSource and syntax to use it is the same. I have about 170 radio stations on my server and only 4 of them using ICECAST. With OkHttpExtension this problem is gone and they play perfectly.

**Comment by ChernyshovYuriy** [comment]
Hi,
I know that issue is closed but I don't want to open new one with the same description.
Using ExoPlayer and Widevine drm, the same behaviour can occur. Hopefully not so often but it can happened with 1 - 2 hour of streaming. Some times it can happened in about 10 - 15 minutes (rarely).
I traced source code and found that execution is looping in
`void doSomeWork() throws ExoPlaybackException`
and in particular
`boolean rendererReadyOrEnded(TrackRenderer renderer)`
At least for **MediaCodecTrackRenderer** it returns **false**.
And here is why.
`long rendererBufferedPositionUs = renderer.getBufferedPositionUs();`
returns **Long.MIN_VALUE** value constantly.

Then, within the same method
`void doSomeWork() throws ExoPlaybackException`
`MediaCodecAudioTrackRenderer` returns **false** (not ready) as well.
And here is why (I traced **MediaCodecTrackRenderer** for the **MediaCodecAudioTrackRenderer**):
```javascript
@Override
protected boolean isReady() {
    boolean isWithin = isWithinHotswapPeriod();
    aLog().d("   IsReady:" + waitingForKeys + " " + sourceState + " " + outputIndex + " " + isWithin);
    return format != null && !waitingForKeys
        && (sourceState != SOURCE_STATE_NOT_READY || outputIndex >= 0 || isWithin);
}
```
and it returns:
`IsReady:false 0 -1 false`

In general, I am concerned about returning **Long.MIN_VALUE**.

Does it helped a little?
